build.assets/tooling/cmd/tag-release/main.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	repos, err := scanForSubmodules(rootRepo, args.recursive)
+	if err != nil {
+		log.Fatalf("Failed scanning repository: %s", err)
+	}
+
+	if err = tagRepos(repos, args.tag, args.annotation, args.force); err != nil {
+		log.Fatalf("Failed during tagging: %s", err)
+	}
+
+	if args.push {
+		if err = pushTags(repos, args.remote, args.tag, args.force); err != nil {
+			log.Fatalf("Failed during tagging: %s", err)
+		}
+	}
+}
+
+// scanForSubmodules collects all of the submodules that the root repoitory has
+// coinfigured (optionally recursively) and compiles them into a single list
+func scanForSubmodules(root *git.Repository, recursive bool) ([]*git.Repository, error) {
+	queue := []*git.Repository{root}
+	result := []*git.Repository{root}
+	var repo *git.Repository
+
+	for len(queue) > 0 {
+		repo, queue = queue[len(queue)-1], queue[:len(queue)-1]
+		w, err := repo.Worktree()
+		if err != nil {
+			return nil, trace.Wrap(err, "failed fetching worktree")
+		}
+
+		repoSubmodules, err := w.Submodules()
+		if err != nil {
+			return nil, trace.Wrap(err, "failed fetching submodules")
+		}
+
+		for _, m := range repoSubmodules {
+			log.Printf("Found submodule %s at %s", m.Config().Name, w.Filesystem.Join(w.Filesystem.Root(), m.Config().Path))
+
+			submoduleRepo, err := m.Repository()
+			if err != nil {
+				return nil, trace.Wrap(err, "Failed fetching submodule repository")
+			}
+
+			result = append(result, submoduleRepo)
+
+			if recursive {
+				queue = append(queue, submoduleRepo)
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// tagRepos tage all repos in the supplied list with a given tag and optional
+// annotation.
+//
+// The default behaviour is to halt at the first error, but if force is `true`,
+// then `tagRepos` will ignore some errors and try to continue. Ignored errors
+// include:
+//   - An existing tag with the same name
+//   - Failure to extract repo metadata
+//   - failure to tag a single repo.
+//
+// In the case of an existing tag, `force` will cause the existing tag to be
+// deleted and re-created. in all other cases, `tagRepos` will give up on the
+// current repsitory and move on to the next one.
+func tagRepos(repos []*git.Repository, tag, annotation string, force bool) error {
+	var opts *git.CreateTagOptions
+	if annotation != "" {
+		opts = &git.CreateTagOptions{
+			Message: annotation,
+		}
+	}
+
+	for _, repo := range repos {
+		w, err := repo.Worktree()
+		if err != nil {
+			if !force {
+				return trace.Wrap(err, "Failed extracting worktree")
+			}
+			log.Printf("WARNING: failed extracting woirktree")
+			continue
+		}
+
+		log.Printf("Tagging %s", w.Filesystem.Root())
+
+		tagExists, err := statTag(repo, tag)
+		if err != nil {
+			return trace.Wrap(err, "failed enumerating tags")
+		}
+
+		if tagExists {
+			if !force {
+				return trace.BadParameter("tag already exists")
+			}
+			log.Printf("WARNING: tag \"%s\" already exists. Deleting.", tag)
```

Comment:
`%q` quotes the string for you
```suggestion
			log.Printf("WARNING: tag %q already exists. Deleting.", tag)
```